### ISSUE TITLE
Question about grafana 2.0 custom query

### ISSUE THREAD

**Issue by gsaray101** [grafana/grafana#1865]
Hi Torkel:

I am using Opentsdb as the backend storage and Graphite webapp for the functions and building trees so that I can organize my metrics by application, type, etc.

The problem I am having is that not all of my apps are tagged due to lack of CMDB etc. Also, I really dont want to tag each metric by app since there are 10s of thousands of servers and metrics. Also updating tags in OpenTSDB is currently not possible if tag information changes. For example lets's say server=webserver101 application=web_pod1, at some point in production, this could change to server=webserver101 application=web_pod2 and for me update opentsdb with this information is very difficult, nearly impossible if there are lots of changes going on.

I see in grafana 2.0 that you have a new data source "Generic SQL type".

If I maintain applicatiion to host mapping in a sql database (Oracle table), which will be very easy and straight forward to udpate, any server to application mapping.

first I will get my hosts for that application from the sql database then I will query graphite to get the metrics data for this application.

Not sure I am clear here but, would it be possible to to include 2 data sources in the same panel as an option? 1st I could call the sql to build first part of the query then 2nd given the first part, I will populate the rest of the query. Any thoughts?

**Comment by torkelo** [comment]
SQL data source (when it is implemented), will allow you to query sql for metrics, like you can for graphite or OpenTSDB. 

I think you should open a change request issue for OpenTSDB so they can improve the use case when you want to change tags.

**Comment by gsaray101** [comment]
How much work will be to use 2 data sources to build a query? Above mentioned is one use case but I am sure there will be others to combine data from different sources?

**Comment by torkelo** [comment]
Well there is one thing to combine two data sources on one graph, what you are talking about is more using one data source to build a query to the second

**Comment by gsaray101** [comment]
yes, I would love to be able to build one query from 2 different data sources.

**Comment by torkelo** [comment]
well, then you will have to fork grafana, because that would require a lot of custom code to pipe the result of one query into building a query for another data source, sounds like very big customization of how to construct queries, but Grafana is open source so everything is possible, just modify the source :) 

So do not expect a feature like this, it sounds incredibly complex. But you can always do something like this by writing an external program that queries SQL and builds dashboards and graph queries based on the SQL results and then updates dashboards in grafana via HTTP api

**Comment by gsaray101** [comment]
Torkel, 

there are two different systems. 1)application, server mapping (sql table), which changes all the time 2. metrics db (opentsdb that only has host info)

1st I have to query the sql table to get host names for a given application. This has to be user driven from the ui, user need to pick an application among several hundred,
Then given hosts, I can query the opentsdb to get other data.

would you mind Elaborating on how would you go about doing this?

**Comment by torkelo** [comment]
Write a SQL data source for grafana that implements the templating query variable interface (metricFind), create two template variables, $app,  $server   (query type), that query the sql for you mapping table. Use the variables in you OpenTSDB query

**Comment by gsaray101** [comment]
sorry to drag-on with this, can I use the template variables with graphite?

**Comment by torkelo** [comment]
sure

**Comment by gsaray101** [comment]
do you have a demo how you do this?

**Comment by torkelo** [comment]
http://play.grafana.org/dashboard/db/templated-graphs-nested

**Comment by gsaray101** [comment]
Torkel,

you got me exited with this. Just to re-iterate this so that I am clear:

1. I will create a $AppName template name what will to to sql and retrieve the app name, host name
2. given an application (lets say web), from the drop down, I will get server list that belong to the web application
3. In my query, can pick $AppName, $ServerName template variables, the rest will come from graphite tree?

**Comment by torkelo** [comment]
@gsaray101 
1) There is no SQL data source yet, so you would have to write on,
2) yes,
3) not sure I understand you but I think yes, but did you not use OpenTSDB?

**Comment by gsaray101** [comment]
1) when you will have the sql data source implemented?
3) I use OpenTSDB as the backend storage and graphite web-app for the the front end. I point grafana to graphite to take advantage of the existing functions that graphite provides.

**Comment by torkelo** [comment]
1) I do not know :) 
3) ah, I see

**Comment by gsaray101** [comment]
1) I do see see the sql data source in grafana 2.0. I guess that's not working?

**Comment by torkelo** [comment]
it was only a test, removed in 2.0.2

**Comment by gsaray101** [comment]
Torkel,

Assuming that sql data source is implemented.

my sql table has the info like this:

Application   Server
Web             j2kweb01
Web             j2kweb02
Tomcat       j2ktk01

Graphite:

My metrics like this:

Source.datacenter.Guest.Reourse_Utilization.server
vSphereGuest.dc1.Guest.Resource_Utilization.j2kweb01

I guess I am really trying to find out i that in graphite there is not application part of the metric string. Is it possible to create a template for the app from sql query that will pick the servers from the graphite. Use the output of the sql based template to build the graphite query?

**Comment by torkelo** [comment]
No idea, maybe. Not sure I understand the flow of queries & templates :)

**Comment by gsaray101** [comment]
sorry, let me try again. Application name is not in the graphite metrics string. Only servers and everything else.

there is another sql table that application mapped to servers. If I retrieve the application specific servers to $server variable, I assume $server is a global variable in in javascript which can be used in the query builder?

**Comment by torkelo** [comment]
yes, any all variable values can be used in queries
